### ISSUE TITLE
Saga exception when shutting down a process: System.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.  

### ISSUE THREAD

**Issue by xhafan** [rebus-org/Rebus#1186]
I use `EnforceExclusiveAccess` for saga, and experienced this exception:
```
System.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.
   at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)
   at Rebus.Sagas.Exclusive.SemaphoreSlimExclusiveSagaAccessIncomingStep.ReleaseLockAsync(Int32 lockId)
   at Rebus.Sagas.Exclusive.EnforceExclusiveSagaAccessIncomingStepBase.ReleaseLocks(Int32[] lockIds)
   at Rebus.Sagas.Exclusive.EnforceExclusiveSagaAccessIncomingStepBase.Process(IncomingStepContext context, Func`1 next)
   at Rebus.Pipeline.Receive.ActivateHandlersStep.Process(IncomingStepContext context, Func`1 next)
...
```
What happened was that the process was being shut down, and the semaphore was not entered, and the wait to enter the semaphore was cancelled, and because the [lock release is in the finally block](https://github.com/rebus-org/Rebus/blob/7344f05e1a2a521698370e068e039e114020a940/Rebus/Sagas/Exclusive/EnforceExclusiveSagaAccessIncomingStepBase.cs#L70), it tries to release the semaphore which was not entered. Basically the same issue as described in [this SO answer](https://stackoverflow.com/a/65535786/379279).

**Comment by xhafan** [comment]
I have a fix which is not 100% correct in my opinion. [EnforceExclusiveSagaAccessIncomingStepBase](https://github.com/rebus-org/Rebus/blob/7344f05e1a2a521698370e068e039e114020a940/Rebus/Sagas/Exclusive/EnforceExclusiveSagaAccessIncomingStepBase.cs#L63):
```
var areLocksAcquired = false;
try
{
    await WaitForLocks(locksToObtain);
    areLocksAcquired = true;

    await next();
}
finally
{
    if (areLocksAcquired)
    {
        await ReleaseLocks(locksToObtain);
    }
}
```
The problem with this fix is that there are potentially multiple locks acquired (=semaphores entered), and only some of them might be acquired, and some of them not, so it would be probably better to have a list of "so far acquired locks" and release only these.

**Comment by mookid8000** [comment]
You fixed it 😺 and it's out in Rebus 8.7.1 on NuGet.org now
